Java8 implementations of collection initializers

danhermann opened the pull request:
Adds Java8 implementations of the `Set::of`, `Set::copyOf`, `List::of`, `List::copyOf`, `Map::of`, `Map::ofEntries`, and `Map::copyOf` static collection methods. Includes classes that delegate to the standard library methods that are available in Java11 for packaging in a multi-release jar.

jasontedor commented on libs/core/src/main/java9/org/elasticsearch/common/collect/List.java:
```diff
@@ -0,0 +1,35 @@
+    @SafeVarargs
+    @SuppressWarnings("varargs")
+    public static <T> java.util.List<T> of(T... entries) {
+        return java.util.List.of(entries);
```
The JDK 9 collections have specializations for empty, one, and two-element lists; I think we could have those too.

jasontedor commented on libs/core/src/main/java9/org/elasticsearch/common/collect/Map.java:
```diff
@@ -0,0 +1,98 @@
+
+package org.elasticsearch.common.collect;
+
+public class Map {
```
We probably want the `Map.ofEntries` method too, for both Java 8 and Java 9+.

jasontedor replied:
Additionally, `Map.copyOf`.

jasontedor commented on libs/core/src/main/java/org/elasticsearch/common/collect/List.java:
```diff
@@ -0,0 +1,38 @@
+     */
+    @SafeVarargs
+    @SuppressWarnings("varargs")
+    public static <T> java.util.List<T> of(T... entries) {
```
I think that we will want a specialization for the empty list (`List.of()` -> `Collections.emptyList()`);

jasontedor commented on libs/core/src/main/java/org/elasticsearch/common/collect/List.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.Arrays;
+import java.util.Collections;
+
+public class List {
```
I think that we will want `List.copyOf` too.

jasontedor commented on libs/core/src/main/java/org/elasticsearch/common/collect/Map.java:
```diff
@@ -0,0 +1,113 @@
+import java.util.Collections;
+import java.util.HashMap;
+
+public class Map {
```
We will want `Map.copyOf` and `Map.ofEntries` too.

jasontedor commented on libs/core/src/main/java/org/elasticsearch/common/collect/Set.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.Collections;
+import java.util.HashSet;
+
+public class Set {
```
We probably want `Set.copyOf`.

jasontedor commented on libs/core/src/main/java/org/elasticsearch/common/collect/Set.java:
```diff
@@ -0,0 +1,39 @@
+     */
+    @SafeVarargs
+    @SuppressWarnings("varargs")
+    public static <T> java.util.Set<T> of(T... entries) {
```
I think we will want a specialization for the empty set (`Set.of()` -> `Collections.emptySet()`).

jasontedor commented on libs/core/src/main/java9/org/elasticsearch/common/collect/Set.java:
```diff
@@ -0,0 +1,36 @@
+
+package org.elasticsearch.common.collect;
+
+public class Set {
```
I think we will want `Set.copyOf`.

jasontedor commented on libs/core/src/main/java9/org/elasticsearch/common/collect/Set.java:
```diff
@@ -0,0 +1,36 @@
+     */
+    @SafeVarargs
+    @SuppressWarnings("varargs")
+    public static <T> java.util.Set<T> of(T... entries) {
```
I think we will want specializations for empty, one, and two-element sets.

danhermann commented:
Thanks for the review, @jasontedor. I added the `copyOf` and `ofEntries` methods for those classes. I also added zero and one-element overloads since they have optimized Java8 implementations through the `Collections.empty*` and `Collections.singleton*` methods. Is there value in adding two-element overloads since, as far as I can see, those would just delegate to the varargs method?

jasontedor commented:
>Is there value in adding two-element overloads since, as far as I can see, those would just delegate to the varargs method?

The one-element and two-element specializations of `List` and `Set` share the same backing implementation, which merely has two fields, instead of a backing array. So they're indeed an optimization.

danhermann commented:
@elasticmachine update branch

jasontedor commented on libs/core/build.gradle:
```diff
@@ -44,13 +44,13 @@ if (!isEclipse) {
   }
 
   compileJava9Java {
```
This should be `compileJava11Java`

jasontedor replied:
We also need to change the source sets above, etc.

Here's what I suggest, let's make changing this to Java 11 a separate change, so it can be reviewed separately. Can you open another pull request for this? Then we can merge that one, and rebase this PR on top of that.

danhermann replied:
@jasontedor, I've rebased this PR now that the separate PR to update the MRJARs to Java11 has been merged.

jasontedor commented on libs/core/build.gradle:
```diff
@@ -44,13 +44,13 @@ if (!isEclipse) {
+    targetCompatibility = 11
   }
 
   forbiddenApisJava9 {
```
This should be `forbiddenApisJava11`

danhermann commented:
@elasticmachine update branch

DaveCTurner commented:
I wonder whether we should also have this in `master`. This work is an enormously helpful step forwards in simplifying backports from `master` to `7.x`, but we still cannot use an unqualified `List.of()` etc in cases where we already `import java.util.List;` since this is a different `List`, and that means a backport will need to change all the call sites to being a fully-qualified `org.elasticsearch.common.collect.List.of();`. IMO it'd help further if we could use the fully-qualified version in both `master` and `7.x`. Once `7.x` is no more it would be a relatively simple refactoring to clean this up.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Core)
